### PR TITLE
Track federated calls

### PR DESCRIPTION
**Pull request by isovector**
As part of gracefully handling offline backends, we want to make sure we have all of our bases covered when it comes to tracking what changes need to happen on the federation side. But it's unclear which endpoints call which federated endpoints. 

This PR tracks federated calls in the type system, ensuring all code paths (including endpoints) document which federated calls they make. For the time being, these constraints get eliminated by orphan instances at the very edge of the APIs, but #2950 follows up on this change, adding a servant combinator for better dissolution of the new `CallsFed` constraints, ensuring they get documented in swagger. 

The relevant change is in `Wire.API.Federation.API`, adding `CallsFed`. Everything else was done automatically in a tireless cycle of asking HLS to add a necessary constraint until the project compiled again.

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

**Comment by isovector** [comment]
@julialongtin @fisx

File: libs/wire-api-federation/src/Wire/API/Federation/API.hs
Comment (mdimjasevic) [inline libs/wire-api-federation/src/Wire/API/Federation/API.hs:54]: I'm yet to look at this thoroughly, but isn't `CallsFed comp name` the same as

```haskell
type CallsFed comp name = forall e. HasFedEndpoint comp e name
```

that is, so long as there exists a federation API endpoint `e` by name `name` in component `comp` (Galley, Brig or Cargohold), `CallsFed comp name` holds?

Comment (pcapriotti) [reply libs/wire-api-federation/src/Wire/API/Federation/API.hs:54]: No, `forall` is the universal quantifier ∀. So `forall e. X e` means that the constraint `X` holds for all possible values of `e`. What you are probably thinking about here is `exists` aka `∃`, which haskell doesn't have for whatever reason. You can emulate `exists e. X e` using "impredicative encoding"/CPS/Church encoding as follows:
```haskell
data ExistsX = forall e. ExistsX (X e)
```
This means that `ExistsX` is equipped with a constructor of type `ExistsX :: forall e. X e → ExistsX`, which by the adjunction relating `exists` and `forall`, is equivalent to `(exists e. X e) → ExistsX`.

In any case, I don't think this is very realistic. For one, if I'm not mistaken, existentially quantified constraints are simply not supported. Even if they were, I doubt you would get any sort of reasonable type inference out of them, so you'd still have to specify them manually. Haskell is just not the right language for encoding logic in types.

**Comment by supersven** [comment]
Hey @isovector :wave: 

Please fix this linter issue:
```
+ make lint-all

./tools/ormolu.sh -c

ormolu mode: check

language extensions: AllowAmbiguousTypes BangPatterns ConstraintKinds DataKinds DefaultSignatures DerivingStrategies DerivingVia DeriveFunctor DeriveGeneric DeriveLift DeriveTraversable EmptyCase FlexibleContexts FlexibleInstances FunctionalDependencies GADTs InstanceSigs KindSignatures LambdaCase MultiParamTypeClasses MultiWayIf NamedFieldPuns NoImplicitPrelude OverloadedStrings PackageImports PatternSynonyms PolyKinds QuasiQuotes RankNTypes ScopedTypeVariables StandaloneDeriving TupleSections TypeApplications TypeFamilies TypeFamilyDependencies TypeOperators UndecidableInstances ViewPatterns

Checking 1382 file(s)…

libs/api-client/src/Network/Wire/Client/Session.hs...  ok

libs/brig-types/src/Brig/Types/User.hs...  ok

libs/gundeck-types/src/Gundeck/Types/Push/V2.hs...  ok

libs/types-common-aws/src/Util/Test/SQS.hs...  ok

libs/wire-api/test/unit/Test/Wire/API/User/Search.hs...  ok

services/proxy/src/Proxy/Run.hs...  ok

tools/stern/src/Stern/Types.hs...  ok

./tools/hlint.sh -f all -m check

services/brig/src/Brig/Run.hs:123:10-23: Warning: Redundant bracket

Found:

  (HasCallStack)

Perhaps:

  HasCallStack


1 hint
```

If I got this right, the general approach is implemented and discussed in https://github.com/wireapp/wire-server/pull/2950 which is still a *draft*. I'm wondering if we should merge this PR, before we agreed upon the general approach. My spontaneous answer would be: No.

Would you mind if we don't merge this PR now, but wait until https://github.com/wireapp/wire-server/pull/2950 is ready and merge them at the same time? I'm very sorry, but I'm hesitant to merge a PR that doesn't add functionality (as that's delivered with https://github.com/wireapp/wire-server/pull/2950.) Please, correct me if I'm wrong: I haven't had time to take a deep dive into this.

File: libs/wire-api-federation/src/Wire/API/Federation/API.hs
Comment (isovector) [reply libs/wire-api-federation/src/Wire/API/Federation/API.hs:54]: >  so long as there exists a federation API endpoint e by name name in component comp (Galley, Brig or Cargohold), CallsFed comp name holds?

The goal of `CallsFed`  is that it *doesn't have any instances,* ie, it shouldn't be dischargable. This forces all downstream callers to also add the constraint, so that we can guarantee it makes it down to the servant `Server` implementation for the API. I'll add some commentary to that effect.

**Comment by isovector** [comment]
> @supersven  If I got this right, the general approach is implemented and discussed in https://github.com/wireapp/wire-server/pull/2950 which is still a draft. I'm wondering if we should merge this PR, before we agreed upon the general approach.

#2950 is done, just marked as a draft because it depends on this one. We can merge only #2950 if you'd like. My thought was because this PR is so big  it's likely to cause lots of merge conflicts the longer it goes unlanded, but #2950 came out faster than I was expecting. Your call.

**Comment by supersven** [comment]
> https://github.com/wireapp/wire-server/pull/2950 is done, just marked as a draft because it depends on this one. We can merge only https://github.com/wireapp/wire-server/pull/2950 if you'd like. My thought was because this PR is so big it's likely to cause lots of merge conflicts the longer it goes unlanded, but https://github.com/wireapp/wire-server/pull/2950 came out faster than I was expecting. Your call.

Ah, I see. :bulb: I think it would be good to wait until the reviewers of https://github.com/wireapp/wire-server/pull/2950 agree that this is the way to go and the name and structure of `CallsFed` is stable. (Only that. Details may be left aside.)

However, I'll be on vacation during the whole next week. So, please feel free to make other plans with your PR's current reviewer(s).

File: services/brig/src/Brig/API/Internal.hs
Comment (mdimjasevic) [inline services/brig/src/Brig/API/Internal.hs:115]: I'd expect the constraint to be gone by now, i.e., exactly at the point where endpoints are served. Has the constraint made it to here by mistake?

File: services/brig/src/Brig/API/Internal.hs
Comment (mdimjasevic) [inline services/brig/src/Brig/API/Internal.hs:301]: ditto

File: services/brig/src/Brig/API/Public.hs
Comment (mdimjasevic) [inline services/brig/src/Brig/API/Public.hs:194]: ditto

File: services/brig/src/Brig/Team/API.hs
Comment (mdimjasevic) [inline services/brig/src/Brig/Team/API.hs:107]: Shouldn't the constraint be gone by now?

File: services/cargohold/src/CargoHold/API/Public.hs
Comment (mdimjasevic) [inline services/cargohold/src/CargoHold/API/Public.hs:43]: ditto

File: services/galley/src/Galley/API/Federation.hs
Comment (mdimjasevic) [inline services/galley/src/Galley/API/Federation.hs:112]: Shouldn't the constraints be gone by now?

File: services/galley/src/Galley/API/Internal.hs
Comment (mdimjasevic) [inline services/galley/src/Galley/API/Internal.hs:116]: Can this instance be moved to somewhere further up, e.g., to `Galley.Run`?

File: services/galley/src/Galley/API/Public/Servant.hs
Comment (mdimjasevic) [inline services/galley/src/Galley/API/Public/Servant.hs:51]: Why they haven't dissolved by now?

**Review by mdimjasevic — APPROVED** [review]
Thanks for explaining what I didn't get at the beginning of reviewing!

There's one thing I thought would happen, but I don't see it in PR: shouldn't the `CallsFed` constraint dissolve when you are combining endpoint handlers into Servant `ServerT` in sitemaps (and similarly for wai-routes)?

Overall, this looks good and I'm approving. If I got the thing above wrong, feel free to merge right away.

**Comment by isovector** [comment]
> There's one thing I thought would happen, but I don't see it in PR: shouldn't the CallsFed constraint dissolve when you are combining endpoint handlers into Servant ServerT in sitemaps (and similarly for wai-routes)?

You're right, nice catch. I'll get that cleaned up today.

**Comment by isovector** [comment]
After 30 minutes of looking into this, it's surprisingly hard to do better here. GHC rejects any attempts to automatically solve all `CallsFed` constraints locally without giving an instance. If we give an instance higher up in Brig, we now require a separate module to contain the orphan instance, since we can't give the same orphan instance in the two files that would need it. Seeing as this all is going to get torn out immediately in #2950 when we introduce the servant combinators, I don't think it's worth putting more energy in here.

Instead I'll add the changelog and land this PR, and we can get working on getting #2950 in sooner.

**Comment by isovector** [comment]
@mdimjasevic Unless you disagree with my earlier comment, please merge this when you get a chance (I don't have the necessary permissions.)

**Comment by mdimjasevic** [comment]
> @mdimjasevic Unless you disagree with my earlier comment, please merge this when you get a chance (I don't have the necessary permissions.)

Merged! You can now proceed with rebasing #2950.

**Pull request merged**
